lib/concurrent/timer_task2.rb

```diff
@@ -0,0 +1,324 @@
+    #     block can control its own lifecycle. Necessary since `self` will
+    #     refer to the execution context of the block rather than the running
+    #     `TimerTask`.
+    #   @yieldparam cancellation a cancellation object representing the joined cancellation
+    #     of the timer task and this run's timeout
+    #
+    #   @return [TimerTask] the new `TimerTask`
+    def initialize(opts = {}, &task)
+      raise ArgumentError.new('no block given') unless block_given?
+      raise ArgumentError.new('reschedule must be either :before or :after') unless [nil, :before, :after].include?(opts[:reschedule])
+      @ExecutionInterval = AtomicReference.new nil
+      @TimeoutInterval = AtomicReference.new nil
+      @Cancellation = AtomicReference.new nil
+
+      self.execution_interval = opts[:execution] || opts[:execution_interval] || EXECUTION_INTERVAL
+      self.timeout_interval = opts[:timeout] || opts[:timeout_interval] || TIMEOUT_INTERVAL
+
+      @Channel = opts[:channel]
+
+      @Task = task
+      @Executor = SafeTaskExecutor.new(task)
+      @Reschedule = opts[:reschedule] || :after
+      @RunNow = opts[:now] || opts[:run_now]
+    end
+
+    # Create a new `TimerTask2` with the same configuration
+    #
+    # @example
+    #   task = Concurrent::TimerTask2.execute(execution_interval: 10) { puts "Hello" }
+    #   task2 = task2.dup
+    def dup(opts = {}, &task)
+      options = {
+        :channel => @Channel,
+        :execution => execution_interval,
+        :now => @RunNow,
+        :reschedule => @Reschedule,
+        :timeout => timeout_interval
+      }
+      self.class.new(options.merge(opts), &(task || @Task))
+    end
+
+    # Create and execute a new `TimerTask2`.
+    #
+    # @!macro timer_task_initializ
+    #
+    # @example
+    #   task = Concurrent::TimerTask2.execute(execution_interval: 10) { print "Hello World\n" }
+    #   task.running? #=> true
+    def self.execute(*args, &task)
+      self.new(*args, &task).execute
+    end
+
+    # Execute a previously created `TimerTask2`.
+    #
+    # @return [TimerTask2] a reference to `self`
+    #
+    # @example Instance and execute in separate steps
+    #   task = Concurrent::TimerTask2.new(execution_interval: 10) { print "Hello World\n" }
+    #   task.running? #=> false
+    #   task.execute
+    #   task.running? #=> true
+    #
+    # @example Instance and execute in one line
+    #   task = Concurrent::TimerTask2.new(execution_interval: 10) { print "Hello World\n" }.execute
+    #   task.running? #=> true
+    def execute
+      execute?
+      self
+    end
+
+    def execute?
+      return false if running?
+      raise IllegalOperationError.new('TimerTask2 cannot be resumed after it was stopped') if @Executor.nil?
+      @Cancellation.set(Cancellation.new)
+      execute_task cancellation, @RunNow
+      true
+    end
+
+    def shutdown
+      # TODO: Implement soft shutdown
+      kill
+    end
```

Comment:
How exactly should soft shutdown behave?

Reply:
Hm I think there are actually 3 levels to consider.
- stop: finish task if running, do not schedule a new tasks
- shutdown: also cancel running task, do not schedule a new tasks
- kill: kill running tasks by killing it's thread, do not schedule a new tasks
What did the old implementation support? 
For the new one I think all 3 make sense, could you implement them? Please free to experiment with the names or API for this, I am not sure myself what will be the best option.

Reply:
Not sure if we can do `kill` in a sensible way, will take a look what we can do

Reply:
The kill will always be dangerous. However, has to be there since the other methods can fail :/ It should be only used as a last resort when the previous methods failed to finish. If you run out of time just put raise NotImplemented there, or we could also meet for coffee and figure something together.